common/thorhelper/thorread.cpp

```diff
@@ -25,33 +25,184 @@
+    : mode(_mode), format(_format), actualCrc(_actualCrc), actualMeta(&_actual), expectedCrc(_expectedCrc), expectedMeta(&_expected), projectedCrc(_projectedCrc), projectedMeta(&_output), options(_options)
+    {}
+
+    virtual const char * queryFormat() const override { return format; }
+    virtual unsigned getActualCrc() const override { return actualCrc; }
+    virtual unsigned getExpectedCrc() const override { return expectedCrc; }
+    virtual unsigned getProjectedCrc() const override { return projectedCrc; }
+    virtual IOutputMetaData * queryActualMeta() const override { return actualMeta; }
+    virtual IOutputMetaData * queryExpectedMeta() const override{ return expectedMeta; }
+    virtual IOutputMetaData * queryProjectedMeta() const override{ return projectedMeta; }
+    virtual const IPropertyTree * queryOptions() const override { return options; }
+    virtual RecordTranslationMode queryTranslationMode() const override { return mode; }
+
+    virtual const IDynamicTransform * queryTranslator() const override
+    {
+        ensureTranslators();
+        return translator;
+    }
+    virtual const IKeyTranslator *queryKeyedTranslator() const override
+    {
+        ensureTranslators();
+        return keyedTranslator;
+    }
+
+    virtual bool matches(const IDiskReadMapping * other) const
+    {
+        return mode == other->queryTranslationMode() && streq(format, other->queryFormat()) &&
+                ((actualCrc && actualCrc == other->getActualCrc()) || (actualMeta == other->queryActualMeta())) &&
+                ((expectedCrc && expectedCrc == other->getExpectedCrc()) || (expectedMeta == other->queryExpectedMeta())) &&
+                ((projectedCrc && projectedCrc == other->getProjectedCrc()) || (projectedMeta == other->queryProjectedMeta())) &&
+                areMatchingPTrees(options, other->queryOptions());
+    }
+
+    virtual bool expectedMatchesProjected() const
+    {
+        return (expectedCrc && expectedCrc == projectedCrc) || (expectedMeta == projectedMeta);
+    }
+
+protected:
+    void ensureTranslators() const;
+
+protected:
+    RecordTranslationMode mode;
+    mutable std::atomic<bool> checkedTranslators = { false };
+    StringAttr format;
+    unsigned actualCrc;
+    unsigned expectedCrc;
+    unsigned projectedCrc;
+    Linked<IOutputMetaData> actualMeta;
+    Linked<IOutputMetaData> expectedMeta;
+    Linked<IOutputMetaData> projectedMeta;
+    Linked<const IPropertyTree> options;
+    mutable Owned<const IDynamicTransform> translator;
+    mutable Owned<const IKeyTranslator> keyedTranslator;
+    mutable SpinLock translatorLock; // use a spin lock since almost certainly not going to contend
+};
+
+void DiskReadMapping::ensureTranslators() const
+{
+    if (checkedTranslators.load())
+        return;
+    SpinBlock block(translatorLock);
+    if (checkedTranslators.load())
```

Comment:
am curious why use load() here and above?, and not just:
```
if (checkedTranslators)
```

Reply:
Because I wanted to make it clear that it was an atomic - and therefore wouldn't suffer from the problem of not using an atomic in this situation.